File: guides/micronaut-security-session-database-authentication/java/src/main/java/example/micronaut/BCryptPasswordEncoderService.java
package example.micronaut;

import io.micronaut.core.annotation.NonNull;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import jakarta.inject.Singleton;
import jakarta.validation.constraints.NotBlank;

@Singleton // <1>
class BCryptPasswordEncoderService implements PasswordEncoder {

    org.springframework.security.crypto.password.PasswordEncoder delegate = new BCryptPasswordEncoder();

    public String encode(@NotBlank @NonNull String rawPassword) {
        return delegate.encode(rawPassword);
    }

    @Override
    public boolean matches(@NotBlank @NonNull String rawPassword,
                           @NotBlank @NonNull String encodedPassword) {
        return delegate.matches(rawPassword, encodedPassword);
    }
}

File: guides/micronaut-database-authentication-provider/python/src/example/micronaut/b_crypt_password_encoder_service.py
from jakarta.inject import Singleton
from org.springframework.security.crypto.bcrypt import BCryptPasswordEncoder

from .password_encoder import PasswordEncoder


@Singleton  # <1>
class BCryptPasswordEncoderService(PasswordEncoder):
    def __init__(self):
        self.delegate = BCryptPasswordEncoder()

    def encode(self, raw_password: str) -> str:
        return self.delegate.encode(raw_password)

    def matches(self, raw_password: str, encoded_password: str) -> bool:
        return self.delegate.matches(raw_password, encoded_password)
